### PR TITLE
process_data(), cgi_param_multiple(), test.c

### PR DESCRIPTION
**Pull request by DFrostByte**
In function 'process_data()': 

Removed unnecessary code that tests for escaped characters.
Simplified dynamic string allocation by using strndup.
Improved name and value extraction by catering for missing values.
Generally refactored resulting in better readability.

### CODE REVIEW

File: src/session.c [hunk 1]
```diff
@@ -107,7 +107,8 @@ const char *session_error_message[] = {
 // cgi.c
 extern int headers_initialized;
 extern void libcgi_error(int error_code, const char *msg, ...);
-extern formvars *process_data(char *query, formvars **start, formvars **last, const char delim, const char sep);
+extern formvars *process_data(char *query, formvars **start, formvars **last,
+                              const char sep_value, const char sep_name);
```
Comment (LeSpocky) [inline src/session.c]: Declaration and definition have different signatures. One is 

`formvars *process_data(const char *query, formvars **start, formvars **last, const char sep_value, const char sep_name)`

while the other is

`formvars *process_data(char *query, formvars **start, formvars **last, const char sep_value, const char sep_name)`.

File: src/cgi.c [hunk 2]
```diff
@@ -82,79 +82,60 @@ extern formvars *cookie_end;
...
-		if (!data)
+	while (*query)
+	{
+		/* allocate and initialise memory for new formvars */
+		item = (formvars *)malloc(sizeof(formvars));
+		if (!item)
 			libcgi_error(E_MEMORY, "%s, line %s", __FILE__, __LINE__);
 
-		memset(data, 0, sizeof(formvars));
+		memset(item, 0, sizeof(formvars));
```
Comment (LeSpocky) [inline src/cgi.c]: maybe omit this and use calloc in line 100 instead?

File: src/cgi.c [hunk 3]
```diff
@@ -82,79 +82,60 @@ extern formvars *cookie_end;
...
```
Comment (LeSpocky) [inline src/cgi.c]: maybe omit this and use calloc in line 100 instead?

Comment (DFrostByte) [inline src/cgi.c]: Ah yes! I had intended to do that initially.

File: src/session.c (hunk 1 above)
Comment (DFrostByte) [inline src/session.c]: Well spotted.

**Comment by DFrostByte** [comment]
I have added commits fixing the mismatching prototypes LeSpocky pointed out, as well as substituting malloc with calloc, as suggested.

I have now also improved performance in cgi_param_multiple() by not beginning the search from formvars_start each time the function is called.

I'm now about to add test.c which contains code I'm writing to help ensure that the functions are performing as intended.

**Comment by LeSpocky** [comment]
Meanwhile @rafaelsteil gave me push access and I would try to further maintain libcgi. However this would require at least a little testing and review from my side which needs a little time. For this pull request I'd like to have it split up. You already put a lot of things in here and I would like to review them separatly. Are you familiar with git to keep your work if I just take the stuff into master piece by piece?

**Comment by DFrostByte** [comment]
Good news. I had thought about suggesting that to @rafaelsteil.

Admittedly, I felt that I was perhaps making too many changes in one request. The problem is that after having already made the pull request, any further commits are being added automatically. I'm not sure what the method is for only submitting x changes; should I be branching?

I'm happy for you to review my changes however you like. I'm not an experienced git user, so perhaps you could elaborate on what you'd like me to do.

**Comment by LeSpocky** [comment]
GitHub adds all following commits of a branch to an existing pull request until it's closed. To avoid this you have to make different branches per pull request. (However you can use this feature intentionally for e.g. correct something from the discussion.)

Regarding Git and workflow I would recommend reading the chapter »Distributed Git - Distributed Workflows« in the Pro Git book: http://git-scm.com/book/en/v2/Distributed-Git-Distributed-Workflows

GitHub is a little special because of the pull request feature and the integration in the website, so I would go with the topic branch approach. Personally I made good experience with keeping the master branch of all forks clean and without personal changes, so I suggest you make one or more topic branches starting from now. Maybe we should leave this pull request as is then, I'll merge it somehow and we start with topic branches afterwards. I don't want to make it too complicated for a tiny project like this so I guess one master branch and topic branches from everybody wanting to contribute would be sufficient.

It's great to have contributions on this. Let's see how far we get. :-)

**Comment by DFrostByte** [comment]
Sounds good to me. I'll be sure not make further commits to the master branch.

**Comment by DFrostByte** [comment]
I have now created separate pull requests for each of these patches.

**Pull request closed without merging**